torchdp/compute_dp_sgd_privacy.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/env python3
+"""
+Both `apply_dp_sgd_analysis()` and `get_privacy_spent()` functions
+ are based on Google's TF Privacy:
+ https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/analysis/compute_dp_sgd_privacy.py
+
+Command-line script for computing privacy of a model trained with DP-SGD.
+The script applies the RDP accountant to estimate privacy budget of an iterated
+ Sampled Gaussian Mechanism.
+
+Example:
+  compute_dp_sgd_privacy
+    --dataset-size=60000 \
+    --batch-size=256 \
+    --sigma=1.12 \
+    --epochs=60 \
+    --delta=1e-5
+The output states that DP-SGD with these parameters satisfies (2.92, 1e-5)-DP.
+
+In addition, there is a -o or --orders argument for a list of RDP orders
+(usually denoted alpha). All arguments have a default value, including -o.
+"""
+import argparse
+import math
+from privacy_analysis import compute_rdp, get_privacy_spent
+
+def apply_dp_sgd_analysis(q, sigma, steps, orders, delta, printed=True):
+    """Compute and print results of DP-SGD analysis.
+
+    Args:
+        q: the sample rate in SGD.
+        sigma: the noise_multiplier in compute_rdp(), ratio of the
+               standard deviation of the Gaussian noise to
+               the l2-sensitivity of the function to which it is added
+        steps: the number of steps.
+        orders: an array (or a scalar) of RDP orders.
+        printed: boolean, True (by default) to print results on stdout.
+  """
+    rdp = compute_rdp(q, sigma, steps, orders)
+    #  Slight adaptation from TF version, in which
+    # `get_privacy_spent()` has one more arguments and one more element in
+    #  returned tuple, because it can also compute delta for a given epsilon
+    #  (and not only compute epsilon for a targeted delta).
+    eps, opt_order = get_privacy_spent(orders, rdp, delta=delta)
+
+    if printed:
+        print(
+            f"DP-SGD with\n\tsampling rate = {100 * q:.3g}% and"
+            f"\n\tnoise_multiplier = {sigma}"
+            f"\n\titerated over {steps} steps\n  satisfies "
+            f"differential privacy with\n\teps = {eps:.3g} "
+            f"and\n\tdelta = {delta}."
+            f"\n  The optimal RDP order is {opt_order}."
+        )
+
+        if opt_order == max(orders) or opt_order == min(orders):
+            print(
+                "The privacy estimate is likely to be improved by expanding "
+                "the set of orders."
+            )
+
+    return eps, opt_order
+
+
+def compute_dp_sgd_privacy(n,
+                           batch_size,
+                           noise_multiplier,
+                           epochs,
+                           delta,
+                           orders=None,
+                           printed=True,
+):
+    """Compute epsilon based on the given hyperparameters.
```

Comment:
```suggestion
    """Compute epsilon based on the given parameters.
```

Reply:
Why not, but they are so called in articles to be differentiated from internal models' parameters (adapted by SGD). Done.